extensions/objects/templates/algebraic-expression-editor.component.ts

```diff
@@ -20,102 +20,123 @@
+      this.value);
+    let eventType = (
+      this.deviceInfoService.isMobileUserAgent() &&
+      this.deviceInfoService.hasTouchEvents()) ? 'focus' : 'change';
+    // We need the 'focus' event while using the on screen keyboard (only
+    // for touch-based devices) to capture input from user and the 'change'
+    // event while using the normal keyboard.
+    Guppy.event(eventType, (focusObj) => {
+      if (!focusObj.focused) {
+        this.isCurrentAnswerValid();
+      }
+      const activeGuppyObject = (
+        this.guppyInitializationService.findActiveGuppyObject());
+      if (activeGuppyObject !== undefined) {
+        this.hasBeenTouched = true;
+        this.currentValue = activeGuppyObject.guppyInstance.asciimath();
+        if (eventType === 'change') {
+          // Need to manually trigger the digest cycle to make any
+          // 'watchers' aware of changes in answer.
+          this.currentValue = activeGuppyObject.guppyInstance.asciimath();
+          this.isCurrentAnswerValid();
         }
-        ctrl.currentValue = ctrl.value;
-        GuppyConfigurationService.init();
-        GuppyInitializationService.init(
-          'guppy-div-creator',
-          MATH_INTERACTION_PLACEHOLDERS.AlgebraicExpressionInput, ctrl.value);
-        let eventType = (
-          DeviceInfoService.isMobileUserAgent() &&
-          DeviceInfoService.hasTouchEvents()) ? 'focus' : 'change';
-        // We need the 'focus' event while using the on screen keyboard (only
-        // for touch-based devices) to capture input from user and the 'change'
-        // event while using the normal keyboard.
-        Guppy.event(eventType, (focusObj) => {
-          if (!focusObj.focused) {
-            ctrl.isCurrentAnswerValid();
-          }
-          var activeGuppyObject = (
-            GuppyInitializationService.findActiveGuppyObject());
-          if (activeGuppyObject !== undefined) {
-            ctrl.hasBeenTouched = true;
-            ctrl.currentValue = activeGuppyObject.guppyInstance.asciimath();
-            if (eventType === 'change') {
-              // Need to manually trigger the digest cycle to make any
-              // 'watchers' aware of changes in answer.
-              $scope.$apply();
-            }
-          }
-        });
-        if (eventType !== 'focus') {
-          Guppy.event('focus', (focusObj) => {
-            if (!focusObj.focused) {
-              ctrl.isCurrentAnswerValid();
-            }
-          });
+      }
+    });
+    Guppy.event('change', () => {
+      const activeGuppyObject = (
+        this.guppyInitializationService.findActiveGuppyObject());
+      if (activeGuppyObject) {
+        this.currentValue = activeGuppyObject.guppyInstance.asciimath();
+        this.isCurrentAnswerValid();
+      }
+    });
+    if (eventType !== 'focus') {
+      Guppy.event('focus', (focusObj) => {
+        if (!focusObj.focused) {
+          this.isCurrentAnswerValid();
         }
-      };
+      });
     }
-  ]
```

Comment:
Let me know if I've gotten the logic right because I am a bit confused at the moment. So we have two event types: focus and change.
Case 1: eventType = focus
```
        Guppy.event('focus', (focusObj) => {
          if (!focusObj.focused) {
            ctrl.isCurrentAnswerValid();
          }
          var activeGuppyObject = (
            GuppyInitializationService.findActiveGuppyObject());
          if (activeGuppyObject !== undefined) {
            ctrl.hasBeenTouched = true;
            ctrl.currentValue = activeGuppyObject.guppyInstance.asciimath();
          }
        });
```

Case 2: If event type = 'Change:
```
Guppy.event('change', (focusObj) => {
          if (!focusObj.focused) {
            ctrl.isCurrentAnswerValid();
          }
          var activeGuppyObject = (
            GuppyInitializationService.findActiveGuppyObject());
          if (activeGuppyObject !== undefined) {
            ctrl.hasBeenTouched = true;
            ctrl.currentValue = activeGuppyObject.guppyInstance.asciimath();
            $scope.$apply();
          }
        });
        Guppy.event('focus', (focusObj) => {
          if (!focusObj.focused) {
            ctrl.isCurrentAnswerValid();
          }
        });
```

Reply:
Yup, logically your code seems correct, although we can keep the code blocks combined like they were originally right? Since most of it is common between both cases. But I'm fine if this seems more readable and if you'd like to keep it this way.
Also, you're not triggering the digest cycle in case of `change` event (for ajs, `$scope.apply` was used). I guess this is why the warnings are not showing up.